feat: add http client configuration to websocket

denopink opened the pull request:
implements #66

- Leverages arrangehttp.ClientConfig as the configuration struct and constructor for the websocket's http client.
- Instantiate a new http client for every connection attempt.

denopink committed:
feat: add http client configuration to websocket

closes #66

denopink committed:
Merge branch 'main' into denopink/feat/ws-http-client-config

schmidtw commented on internal/websocket/options.go:
```diff
@@ -218,16 +178,51 @@ func SendTimeout(d time.Duration) Option {
-func ConnectTimeout(d time.Duration) Option {
+// HTTPClient is the configuration for the HTTP client used to retrieve the
+// credentials.
+func HTTPClient(c arrangehttp.ClientConfig) Option {
```
This needs to be an optional Option.  Right now it is required since it creates the http.Client used.  I think this is a mistake & that this should largely just take a `*http.Client` in as input.  Later (after the options are handled) the final `*http.Client` should be created given the need for the `ipv4` vs `ipv6` choice as well as which network interface to attempt to use.

denopink replied:
great point(!), I updated it such that `func HTTPClient(c arrangehttp.ClientConfig)` is actually optional and the http client is created as needed 🙂

schmidtw commented on internal/websocket/ws.go:
```diff
@@ -330,14 +321,11 @@ func (ws *Websocket) dial(ctx context.Context, mode ipMode) (*nhws.Conn, *http.R
 	}
-
+	ws.updateTransportDialContext(mode)
 	conn, resp, err := nhws.Dial(ctx, url,
```
This is the place the the final `*http.Client` should be created.  It should not be stored, just passed to the `Dial()` function.

denopink committed:
chore: update based on pr feedback

- instantiate a new http client for every connection attempt
- instantiate the http client in `func (ws *Websocket) newHTTPClient(mode ipMode) (*http.Client, error)`
- replace `Websocket.client` with `Websocket.httpClientConfig`, `httpClientConfig` being the http client configuration and constructor for `newHTTPClient`

denopink commented on internal/websocket/options.go:
```diff
@@ -218,16 +178,38 @@ func SendTimeout(d time.Duration) Option {
+			// test new client
+			_, err = c.NewClient()
+			if err != nil {
+				return errors.Join(errs, err)
```
typo
```suggestion
				return errors.Join(ErrMisconfiguredWS, err)
```

schmidtw commented on internal/websocket/options.go:
```diff
@@ -218,16 +178,38 @@ func SendTimeout(d time.Duration) Option {
-// the default is 30 seconds.
-func ConnectTimeout(d time.Duration) Option {
+// HTTPClient is the configuration for the HTTP client used for connection attempts.
+func HTTPClient(c arrangehttp.ClientConfig) Option {
```
Instead of passing the `arrangehttp.ClientConfig` pass the `*http.Client` value here.  This is consistent with other similar interfaces & decouples configuration from use a bit more.  Also I don't think we need to boundary check the client since we're being told what someone wants & that will remove some code as well.

denopink committed:
chore: pr update

denopink committed:
chore: update based on pr feedback

denopink committed:
Merge branch 'denopink/feat/ws-http-client-config' of https://github.com/xmidt-org/xmidt-agent into denopink/feat/ws-http-client-config

denopink committed:
chore: typo
